Add generation of CF serializers

beidson opened the pull request:
#### b96d6e41056bce05390a67eaa3075b311a5aa59b
<pre>
Add generation of CF serializers
<a href="https://bugs.webkit.org/show_bug.cgi?id=263800">https://bugs.webkit.org/show_bug.cgi?id=263800</a>

Reviewed by David Kilzer.

A decent chunk of remaining handwritten coders are in ArgumentCodersCF.

The strategy is to add a C++ class to wrap each of the CFTypes we serialize,
and that C++ class&apos;s coders are also generated.

CoreIPCData is a prime candidate of this, in that it can easily wrap a CFDataRef.

This patch teachers the generator to look for types that end in &quot;Ref&quot; and are
followed by a &quot;wrapped by&quot; clause.
It then generates the encode/decode methods for that CF type in terms of the wrapper.

CFDataRef and CFStringRef, and already have suitable wrappers, so we&apos;ll start with them.
This can also be extended to NS* types in the future.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.namespace_and_name):
(SerializedType.cf_wrapper_type):
(one_argument_coder_declaration_cf):
(one_argument_coder_declaration):
(generate_header):
(check_type_members):
(encode_cf_type):
(encode_type):
(decode_cf_type):
(decode_type):
(generate_one_impl):
(generate_serialized_type_info):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;CFFooRef&gt;::encode):
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFFooRef&gt;&gt;::decode):
(IPC::ArgumentCoder&lt;CFBarRef&gt;::encode):
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFBarRef&gt;&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFFooRef&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFBarRef&gt;&gt;::encode):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCData.h:
(WebKit::CoreIPCData::CoreIPCData):
* Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder&lt;CFDataRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFDataRef&gt;&gt;::decode): Deleted.
(IPC::ArgumentCoder&lt;CFStringRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFStringRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269951@main">https://commits.webkit.org/269951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6d379755c128c18a1f87acedf8a60c59edcb00d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24015 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26156 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22121 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24498 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22610 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24259 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20759 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26745 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/24189 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1416 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27911 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21892 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21962 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25700 "Found 3 new API test failures: /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/dom-cache, /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/origin-and-total-storage-ratio, /WebKitGTK/TestWebKitFaviconDatabase:/webkit/WebKitFaviconDatabase/ephemeral (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1352 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/19029 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1371 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5770 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1762 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1703 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in:
```diff
@@ -24,7 +24,7 @@
 
-[WebKitPlatform] class WebKit::CoreIPCData {
+[WebKitPlatform, AdditionalEncoder=StreamConnectionEncoder] class WebKit::CoreIPCData {
     IPC::DataReference get();
```
I needed to add: [NotSerialized](https://github.com/WebKit/WebKit/compare/491a015b27ed49fef99abb6fe466dd71f54f1553..79da2a56ae4a85c4901d09d6e0b00908f4fce05f#diff-c53573c4702dac302bb681f594eb635030e08018baf28929fd51e790959d355fL26) before, as I hit a static assert. Do we want that same check?

beidson replied:
My evidence suggests we do not need that, because I'm running tests just fine with no static assert.

beidson replied:
That said, I'm staring *right at* the generated serializers here...

`void ArgumentCoder<WebKit::CoreIPCData>::encode(Encoder& encoder, const WebKit::CoreIPCData& instance)
{
    static_assert(std::is_same_v<std::remove_cvref_t<decltype(instance.get())>, IPC::DataReference>);

    encoder << instance.get();
}

void ArgumentCoder<WebKit::CoreIPCData>::encode(StreamConnectionEncoder& encoder, const WebKit::CoreIPCData& instance)
{
    static_assert(std::is_same_v<std::remove_cvref_t<decltype(instance.get())>, IPC::DataReference>);

    encoder << instance.get();
}`

And I'm not sure how that's *not* firing

beidson replied:
Ohhhh  that's typing, not size.

So yah, something about this declaration is not putting the size in, so...  *shrug*

ddkilzer reviewed (approved):
r=me

beidson committed:
Add generation of CF serializers
https://bugs.webkit.org/show_bug.cgi?id=263800

Reviewed by David Kilzer.

A decent chunk of remaining handwritten coders are in ArgumentCodersCF.

The strategy is to add a C++ class to wrap each of the CFTypes we serialize,
and that C++ class's coders are also generated.

CoreIPCData is a prime candidate of this, in that it can easily wrap a CFDataRef.

This patch teachers the generator to look for types that end in "Ref" and are
followed by a "wrapped by" clause.
It then generates the encode/decode methods for that CF type in terms of the wrapper.

CFDataRef and CFStringRef, and already have suitable wrappers, so we'll start with them.
This can also be extended to NS* types in the future.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.namespace_and_name):
(SerializedType.cf_wrapper_type):
(one_argument_coder_declaration_cf):
(one_argument_coder_declaration):
(generate_header):
(check_type_members):
(encode_cf_type):
(encode_type):
(decode_cf_type):
(decode_type):
(generate_one_impl):
(generate_serialized_type_info):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<CFFooRef>::encode):
(IPC::ArgumentCoder<RetainPtr<CFFooRef>>::decode):
(IPC::ArgumentCoder<CFBarRef>::encode):
(IPC::ArgumentCoder<RetainPtr<CFBarRef>>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
(IPC::ArgumentCoder<RetainPtr<CFFooRef>>::encode):
(IPC::ArgumentCoder<RetainPtr<CFBarRef>>::encode):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCData.h:
(WebKit::CoreIPCData::CoreIPCData):
* Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<CFDataRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<CFDataRef>>::decode): Deleted.
(IPC::ArgumentCoder<CFStringRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<CFStringRef>>::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCData.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/269951@main
